### PR TITLE
Fixing resolve logic in serve and lint

### PR DESCRIPTION
**Pull request by philsturgeon**
When running serve on a $ref-filled schema it fails to load the target. Instead of shows the same index.html so we dont even get a 404, just a invalid response error.

It turns out that lint was also not resolving correctly, so expect to start seeing more linting errors after updating.

For that reason I'm thinking this will be v0.5.0 Don't wanna break things for people on a minor.

**Todo**

- [ ] Need to add some tests

### CODE REVIEW

File: lib/loader.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
+
+class OpenError extends ExtendableError {}
+class ReadError extends ExtendableError {}
+
+function readFileAsync(filename, encoding) {
```
Comment (philsturgeon) [inline lib/loader.js]: Copied from resolver, but resolver is going away in a future version as we ship the dependency to a new package. Let's just be ok with this for now.

File: lib/loader.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        resolve: true,
+        cache: [],
+        externals: [],
+        externalRefs: {},
+        rewriteRefs: true,
+        origin: source,
+        source: source
+    };
+
+    options.openapi = openapi;
```
Comment (perico) [inline lib/loader.js]: Why not ?
```
const options = {
    resolve: true,
    cache: [],
    externals: [],
    externalRefs: {},
    rewriteRefs: true,
    origin: source,
    source: source,
    openapi : openapi
};
```

File: lint.js [hunk 3]
```diff
@@ -78,45 +46,48 @@ ${colors.reset + error.message}
...
-    });
-};
 
-const command = (file, cmd) => {
+const readOrError = file => {
+    try {
+        return loader.loadSpec(file, { resolve: true });
+    }
+    catch (error) {
+        if (error.name == 'OpenError') {
```
Comment (perico) [inline lint.js]: why don't use `if(error instanceof OpenError)`?
same for ReadError

File: lint.js [hunk 4]
```diff
@@ -78,45 +46,48 @@ ${colors.reset + error.message}
...
-          }
-          main(data, lintResolvedSchema);
-      });
-  }
-
+  validator.validate(options.openapi, options, function(err, options) {
+      if (err) {
+          console.log(colors.red + 'Specification schema is invalid.' + colors.reset);
+          formatSchemaError(err, options.context);
+          process.exit(1);
```
Comment (perico) [inline lint.js]: is it ok to exit here? shouldn't throw a validation error or something and be caught in the main func?

Comment (dangoosby) [reply lint.js]: Agreed. There should be some error output for the exit IMO

**Comment by dangoosby** [comment]
Wrote a minor comment which agreed with perico. Looks great.

File: lib/loader.js (hunk 2 above)
Comment (philsturgeon) [reply lib/loader.js]: Ha yes. This was code that changed over time, thanks for catching it.

File: lint.js (hunk 4 above)
Comment (philsturgeon) [reply lint.js]: Hey, this was confusing as the formatSchemaError was outputting content, so there _is_ always content. I've made those functions return strings, and its a bit more clear now. 

```
          const output = formatSchemaError(err, options.context);
          console.error(output);
          process.exit(1);
```

**Pull request merged**